crates/admin/src/cluster_controller/replicated_loglet_controller.rs

```diff
@@ -0,0 +1,656 @@
+                            loglet_id: ReplicatedLogletId::new(loglet_id),
+                            params: LogletParams::from(
+                                params.serialize().expect("can always serialize params"),
+                            ),
+                            segment_index: SegmentIndex::default(),
+                            node_set,
+                            sequencer: Some(sequencer),
+                        });
+
+                        // todo: what if we previously proposed this action? how do we avoid doing this unnecessarily?
+                        effects.push(LogletEffect::AddLogletSegment);
+                    }
+
+                    Some(target_state) => {
+                        // Check if loglet configuration requires any remediating actions
+                        let mut nodes_config = NodesConfiguration::default(); // todo: this should come from observed state
+                        cluster_state
+                            .healthy_workers
+                            .iter()
+                            .to_owned()
+                            .for_each(|(node_id, _)| {
+                                nodes_config.upsert_node(NodeConfig::new(
+                                    format!("node-{}", node_id),
+                                    cluster_state.healthy_workers[node_id],
+                                    format!("unix:/tmp/my_socket-{}", node_id).parse().unwrap(),
+                                    Role::LogServer.into(),
+                                    LogServerConfig {
+                                        storage_state: cluster_state.log_server_states[node_id],
+                                    },
+                                ));
+                            });
+
+                        // 1. Is the log node set healthy? Can we do better than the existing state?
+                        // Nodes passed into the healthy_log_servers list must meet the storage criteria
+                        // for writable log server, so we intersect with those and check if the result is ok.
+                        let effective_node_set = target_state.node_set.to_effective(&nodes_config);
+                        if effective_node_set.len() < self.config.replication.num_copies() as usize
+                        {
+                            info!(log_id = ?target_state.log_id, "Loglet node set is not healthy, reconfiguration required");
+
+                            let node_set = NodeSet::from_nodes(
+                                cluster_state
+                                    .healthy_log_servers
+                                    .keys()
+                                    .cloned()
+                                    .choose_multiple(
+                                        &mut rand::thread_rng(),
+                                        self.config.replication.num_copies() as usize,
```

Comment:
I think we need to pick the configured nodeset size and not the replication property. Otherwise a single loss of a node will make sealing impossible. If replication property is 2, then the nodeset size would be 3.

Reply:
Yeah! I understood this at the time I wrote it, but I misused the `ReplicationProperty` type here. In config, I mean this to be "preferred maximum replication goal" rather than "hard requirement". But then I muddled up the two different meanings in use 😅 I think what I want ultimately is to implement a new type of `SelectorStrategy` that isn't `Flood` but more like `Optimum(PreferredPlacement)` but I didn't want to go there just yet. I've updated with a simple usize desired replication goal config property which will be in the next iteration.